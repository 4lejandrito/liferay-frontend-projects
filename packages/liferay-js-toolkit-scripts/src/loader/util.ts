/**
 * SPDX-FileCopyrightText: Â© 2020 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

import {FilePath} from 'liferay-js-toolkit-core';
import path from 'path';

import {project} from '../config';

const fileOriginCache: {[index: string]: FileOrigin} = {};
const packageDirCache: {[index: string]: FilePath} = {};

enum FileOrigin {
	SOURCE_PROJECT,
	SOURCE_DEPENDENCY,
	BUILD_PROJECT,
	BUILD_DEPENDENCY,
}

/**
 * Get the name of a module given its file path.
 *
 * @remarks
 * This method assumes that the module file lives in a liferay-npm-bundler
 * project. Otherwise, the return path may be incorrect.
 *
 * @param moduleFilePath the module file path
 * @return the name of the module
 */
export function getModuleName(moduleFilePath: string): string {
	const absModuleFilePosixPath = new FilePath(moduleFilePath).resolve()
		.asPosix;
	const absPkgPosixPath = new FilePath(getPackageDir(moduleFilePath)).asPosix;

	let moduleName: string = absModuleFilePosixPath.substring(
		absPkgPosixPath.length + 1
	);

	moduleName = moduleName.replace(/\.js$/i, '');

	if (getFileOrigin(moduleFilePath) === FileOrigin.SOURCE_PROJECT) {
		for (const source of project.sources) {
			const prefix = source.asPosix.substr(2);

			if (moduleName.startsWith(prefix)) {
				moduleName = moduleName.substring(prefix.length + 1);
				break;
			}
		}
	}

	// eslint-disable-next-line @typescript-eslint/no-var-requires, liferay/no-dynamic-require,
	const pkgJson = require(getPackageJsonPath(moduleFilePath));

	return `${pkgJson.name}@${pkgJson.version}/${moduleName}`;
}

/**
 * Get the full path of the package directory for a given JS module file.
 *
 * @remarks
 * This method assumes that the module file lives in a liferay-npm-bundler
 * project. Otherwise, the return path may be incorrect.
 *
 * @param moduleFilePath the module file path
 * @return the full path to the package directory (with native path
 *         separators)
 */
export function getPackageDir(moduleFilePath: string): string {
	const absModuleFilePosixPath = new FilePath(moduleFilePath).resolve()
		.asPosix;

	let absPkgDir: FilePath = packageDirCache[absModuleFilePosixPath];

	if (absPkgDir) {
		return absPkgDir.asNative;
	}

	const fileOrigin = getFileOrigin(moduleFilePath);

	switch (fileOrigin) {
		case FileOrigin.SOURCE_DEPENDENCY: {
			const absModuleFilePosixPath = new FilePath(
				moduleFilePath
			).resolve().asPosix;

			const absNodeModulesPosixPath = project.dir.join('node_modules')
				.asPosix;

			const relModuleFilePosixPath = absModuleFilePosixPath.substring(
				absNodeModulesPosixPath.length + 1
			);

			const modulePathParts = relModuleFilePosixPath.split('/');

			const pkgDir = relModuleFilePosixPath.startsWith('@')
				? `${modulePathParts[0]}/${modulePathParts[1]}`
				: modulePathParts[0];

			absPkgDir = project.dir.join('node_modules', pkgDir);

			break;
		}

		case FileOrigin.BUILD_DEPENDENCY: {
			const absModuleFilePosixPath = new FilePath(
				moduleFilePath
			).resolve().asPosix;

			const absBuildNodeModulesPosixPath = project.dir.join(
				project.buildDir,
				'node_modules'
			).asPosix;

			const relBuildNodeModulesPosixPath = absModuleFilePosixPath.substring(
				absBuildNodeModulesPosixPath.length + 1
			);

			const modulePathParts = relBuildNodeModulesPosixPath.split('/');

			const pkgDir = modulePathParts[0];

			absPkgDir = project.dir.join(
				project.buildDir,
				'node_modules',
				pkgDir
			);

			break;
		}

		case FileOrigin.SOURCE_PROJECT: {
			absPkgDir = project.dir;

			break;
		}

		case FileOrigin.BUILD_PROJECT: {
			absPkgDir = project.dir.join(project.buildDir);

			break;
		}

		default:
			throw new Error(`Unkown FileOrigin: ${fileOrigin}`);
	}

	packageDirCache[absModuleFilePosixPath] = absPkgDir;

	return absPkgDir.asNative;
}

// TODO: get rid of this function once we support loading config from .js files
/**
 * Replace `${project.name}` and `${project.version}` tokens inside strings of
 * loader options objects.
 */
export function replaceTokens<T>(
	options: T,
	{except}: {except: string[]} = {except: []}
): T {
	const {pkgJson} = project;

	Object.keys(options).forEach((key) => {
		if (typeof options[key] === 'string') {
			if (except.includes(key)) {
				return;
			}

			options[key] = options[key]
				.replace(/\$\{project\.name\}/g, pkgJson['name'])
				.replace(/\$\{project\.version\}/g, pkgJson['version']);
		}
	});

	return options;
}

/**
 * Remove a webpack hash (a hex number surrounded by dots near the end of the
 * file name) if present from a file path.
 */
export function removeWebpackHash(filePath: string): string {
	const filePathParts = filePath.split('.');

	let hashIndex;

	for (hashIndex = filePathParts.length - 1; hashIndex >= 0; hashIndex--) {
		const filePathPart = filePathParts[hashIndex];

		if (filePathPart.match(/^[0-9a-fA-F]+$/)) {
			break;
		}
	}

	if (hashIndex === -1) {
		return filePath;
	}

	filePathParts.splice(hashIndex, 1);

	return filePathParts.join('.');
}

function getFileOrigin(filePath: string): FileOrigin {
	const absModuleFilePosixPath = new FilePath(filePath).resolve().asPosix;

	let fileOrigin: FileOrigin = fileOriginCache[absModuleFilePosixPath];

	if (fileOrigin) {
		return fileOrigin;
	}

	const absBuildNodeModulesPosixPath = project.dir.join(
		project.buildDir,
		'node_modules'
	).asPosix;
	const absBuildPosixPath = project.dir.join(project.buildDir).asPosix;
	const absNodeModulesPosixPath = project.dir.join('node_modules').asPosix;

	if (absModuleFilePosixPath.startsWith(absBuildNodeModulesPosixPath)) {
		fileOrigin = FileOrigin.BUILD_DEPENDENCY;
	} else if (absModuleFilePosixPath.startsWith(absBuildPosixPath)) {
		fileOrigin = FileOrigin.BUILD_PROJECT;
	} else if (absModuleFilePosixPath.startsWith(absNodeModulesPosixPath)) {
		fileOrigin = FileOrigin.SOURCE_DEPENDENCY;
	} else {
		fileOrigin = FileOrigin.SOURCE_PROJECT;
	}

	fileOriginCache[absModuleFilePosixPath] = fileOrigin;

	return fileOrigin;
}

function getPackageJsonPath(moduleFilePath: string): string {
	return path.join(getPackageDir(moduleFilePath), 'package.json');
}
