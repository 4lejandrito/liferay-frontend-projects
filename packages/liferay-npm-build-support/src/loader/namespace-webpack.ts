/**
 * Â© 2017 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

import crypto from 'crypto';

import {
	BundlerLoaderContext,
	BundlerLoaderReturn,
} from 'liferay-npm-build-tools-common/lib/api/loaders';
import project from 'liferay-npm-build-tools-common/lib/project';

export default function({
	content,
	log,
}: BundlerLoaderContext): BundlerLoaderReturn {
	const regexp = /webpackJsonp/g;

	const matches = regexp.exec(content);

	if (!matches) {
		log.info(
			'namespace-webpack',
			`No ocurrences of 'webpackJsonp' found; nothing to be done`
		);

		return content;
	}

	log.info(
		'namespace-webpack',
		`Replaced ${matches.length} occurences of 'webpackJsonp'`
	);

	const hash = crypto.createHash('MD5');

	hash.update(project.pkgJson['name']);
	hash.update(project.pkgJson['version']);

	const uuid = hash
		.digest('base64')
		.replace(/\+/g, '_')
		.replace(/\//g, '_')
		.replace(/=/g, '');

	return content.replace(regexp, `webpackJsonp_${uuid}`);
}
