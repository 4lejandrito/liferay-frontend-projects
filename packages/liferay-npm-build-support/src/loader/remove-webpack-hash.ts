/**
 * Â© 2017 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

import {
	BundlerLoaderContext,
	BundlerLoaderReturn,
} from 'liferay-npm-build-tools-common/lib/api/loaders';

export default function(context: BundlerLoaderContext): BundlerLoaderReturn {
	const {content, filePath, extraArtifacts, log} = context;

	const filePathParts = filePath.split('.');

	if (Number.isNaN(parseInt(filePathParts[filePathParts.length - 2], 16))) {
		log.info(
			'remove-webpack-hash',
			`No webpack hash in filename; nothing to be done`
		);

		return;
	}

	const newFilePath =
		filePathParts.slice(0, filePathParts.length - 2).join('.') +
		'.' +
		filePathParts[filePathParts.length - 1];

	extraArtifacts[newFilePath] = content;
	context.content = undefined;

	log.info('remove-webpack-hash', `Diverting file output to ${newFilePath}`);

	return undefined;
}
