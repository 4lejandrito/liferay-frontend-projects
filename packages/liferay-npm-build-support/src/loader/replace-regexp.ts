/**
 * Â© 2017 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

import {
	BundlerLoaderContext,
	BundlerLoaderReturn,
} from 'liferay-npm-build-tools-common/lib/api/loaders';

export default function(
	{content, log}: BundlerLoaderContext,
	{
		pattern,
		flags = 'g',
		replacement,
	}: {pattern: string; flags: string; replacement: string}
): BundlerLoaderReturn {
	const regexp = new RegExp(pattern, flags);

	const matches = regexp.exec(content);

	if (!matches) {
		log.info(
			'replace-regexp',
			`No occurences of ${pattern} found; nothing to be done`
		);

		return content;
	}

	log.info(
		'replace-regexp',
		`Replaced ${matches.length} occurences of ${pattern}`
	);

	return content.replace(regexp, replacement);
}
