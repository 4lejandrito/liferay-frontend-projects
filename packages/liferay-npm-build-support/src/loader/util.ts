/**
 * Â© 2017 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

import project from 'liferay-npm-build-tools-common/lib/project';

// TODO: get rid of this function once we support loading config from .js files
/**
 * Replace `${project.name}` and `${project.version}` tokens inside strings of
 * loader options objects.
 */
export function replaceTokens<T>(
	options: T,
	{except}: {except: string[]} = {except: []}
): T {
	const {pkgJson} = project;

	Object.keys(options).forEach(key => {
		if (typeof options[key] === 'string') {
			if (except.includes(key)) {
				return;
			}

			options[key] = options[key]
				.replace(/\$\{project\.name\}/g, pkgJson['name'])
				.replace(/\$\{project\.version\}/g, pkgJson['version']);
		}
	});

	return options;
}
