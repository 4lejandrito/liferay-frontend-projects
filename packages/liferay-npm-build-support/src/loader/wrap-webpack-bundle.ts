/**
 * Â© 2017 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

import {
	BundlerLoaderContext,
	BundlerLoaderReturn,
} from 'liferay-npm-build-tools-common/lib/api/loaders';
import FilePath from 'liferay-npm-build-tools-common/lib/file-path';
import * as pkgs from 'liferay-npm-build-tools-common/lib/packages';
import readJsonSync from 'read-json-sync';

export default function(
	{content, filePath, log}: BundlerLoaderContext,
	{removePrefix}: {removePrefix: string}
): BundlerLoaderReturn {
	const moduleName = getModuleName(filePath, removePrefix);

	log.info(
		'wrap-webpack-bundle',
		`Wrapped bundle with module name '${moduleName}'`
	);

	return `
Liferay.Loader.define(
	'${moduleName}',
	['module'],
	function(module) {
		module.exports = function(_LIFERAY_PARAMS_) {
			${content}
		}
	}
);`;
}

function getModuleName(prjRelfilePath: string, removePrefix: string): string {
	const absFile = new FilePath(prjRelfilePath);
	const pkgDir = new FilePath(pkgs.getPackageDir(prjRelfilePath));

	let moduleName: string = pkgDir.relative(absFile).asPosix;

	if (moduleName.toLowerCase().endsWith('.js')) {
		moduleName = moduleName.substring(0, moduleName.length - 3);
	}

	const chunkIndex = moduleName.lastIndexOf('.');

	if (!Number.isNaN(parseInt(moduleName.substring(chunkIndex + 1), 16))) {
		moduleName = moduleName.substring(0, chunkIndex);
	}

	if (moduleName.startsWith(removePrefix)) {
		moduleName = moduleName.substring(removePrefix.length);
	}

	const pkgJson = readJsonSync(pkgDir.join('package.json').asNative);

	return `${pkgJson.name}@${pkgJson.version}/${moduleName}`;
}
