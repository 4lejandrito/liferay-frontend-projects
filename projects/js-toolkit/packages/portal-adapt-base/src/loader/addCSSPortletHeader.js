/**
 * SPDX-FileCopyrightText: Â© 2020 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

const project = require('liferay-npm-build-tools-common/lib/project');

const replaceTokens = require('../util/replaceTokens');

/**
 * A loader that adds the `com.liferay.portlet.header-portlet-css` to the
 * `package.json` file.
 *
 * This loader allows adding a CSS file path to be used by the generated portlet
 * on the fly.
 *
 * It is normally used from adapted projects to inject the framework's generated
 * CSS file into the portlet.
 *
 * @remarks
 *
 * Valid options are:
 *
 *   - css: path to the CSS file to use as `header-portlet-css` property of the portlet
 */
module.exports = function addCSSPortletHeader(context, options) {
	const {content, log} = context;
	const {css} = replaceTokens(project.default, options);

	const pkgJson = JSON.parse(content);

	pkgJson['portlet'] = pkgJson['portlet'] || {};

	if (pkgJson['portlet']['com.liferay.portlet.header-portlet-css']) {
		log.warn(
			'add-css-portlet-header',
			'The package.json has a configured header-portlet-css entry, ' +
				'but it will be overwritten'
		);
	}

	pkgJson['portlet']['com.liferay.portlet.header-portlet-css'] = css;

	log.info('add-css-portlet-header', `Added ${css} as portlet CSS file`);

	return JSON.stringify(pkgJson, null, '\t');
};
