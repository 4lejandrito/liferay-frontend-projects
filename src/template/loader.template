<%= vendor %>

if (typeof AUI === 'undefined') {
    var AUI = {};
}

(function (global, factory) {
    var built = factory();

    /* istanbul ignore else */
    if (typeof module === 'object' && module) {
        module.exports = built;
    }

    /* istanbul ignore next */
    if (typeof define === 'function' && define.amd) {
        define(factory);
    }

    global.Loader = new built();
}(typeof global !== 'undefined' ? global : /* istanbul ignore next */ this, function () {
	'use strict';

	<%= source %>

    return Loader;
}));