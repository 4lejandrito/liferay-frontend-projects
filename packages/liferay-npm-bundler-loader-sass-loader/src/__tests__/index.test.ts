/**
 * SPDX-FileCopyrightText: Â© 2017 Liferay, Inc. <https://liferay.com>
 * SPDX-License-Identifier: LGPL-3.0-or-later
 */

import PluginLogger from 'liferay-js-toolkit-core/lib/plugin-logger';

import loader from '../index';

it('logs results correctly', () => {
	const context = {
		content: '$color: red; .Button { border: 1px solid $color; }',
		filePath: 'file.scss',
		log: new PluginLogger(),
		extraArtifacts: {},
	};

	loader(context, {});

	// This test returns '(from project)' because of yarn hoisting, which makes
	// the loader think that it is retrieving sass from a project (as opposed to
	// itself).
	expect(context.log.messages).toEqual([
		{
			level: 'info',
			source: 'sass-loader',
			things: ['Processed file with sass v1.24.0 (from project)'],
		},
	]);
});

it('correctly generates CSS', () => {
	const context = {
		content: '$color: red; .Button { border: 1px solid $color; }',
		filePath: 'file.scss',
		log: new PluginLogger(),
		extraArtifacts: {},
	};

	const result = loader(context, {});

	expect(result).toEqual(`.Button {
  border: 1px solid red;
}`);

	expect(Object.keys(context.extraArtifacts)).toEqual([]);
});
