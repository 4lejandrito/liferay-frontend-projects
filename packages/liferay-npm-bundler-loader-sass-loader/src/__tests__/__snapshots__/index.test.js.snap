// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`correctly generates CSS 1`] = `
".Button {
  border: 1px solid red;
}"
`;

exports[`logs results correctly 1`] = `
Array [
  Object {
    "level": "info",
    "source": "sass-loader",
    "things": Array [
      "Processed file with sass v1.22.2 (from loader)",
    ],
  },
]
`;
